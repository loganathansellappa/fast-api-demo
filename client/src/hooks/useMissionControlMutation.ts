import { useAxiosClient } from '../context/AxiosClientContext.ts';
import { useMutation, useQueryClient } from 'react-query';
import { missionControlKeys } from '../services/queryKeyFactory.ts';
import { MissionState, MutationType } from '../utils/HelperUtils.ts';

export type MissionControlMutation = {
  missionState?: MissionState;
  mutationType: MutationType;
  url: string;
  description?: string;
  title?: string;
};

export type MissionControlInput = {
  mutationType: MutationType;
  id?: number;
  missionState?: MissionState;
  mission_state?: MissionState;
  description?: string;
  title?: string;
};

type RequestData = {
  url: string;
  method: MutationType;
  mission_state?: MissionState;
  data?: Record<string, unknown>;
};

export function useMissionControlUpdate() {
  const axiosClientService = useAxiosClient();
  const queryClient = useQueryClient();

  const mutation = useMutation({
    mutationFn: (requestData: MissionControlMutation) => {
      const { url, mutationType, missionState, ...data } = requestData;
      const requestObject: RequestData = {
          url: url,
          method: mutationType
      }
      if (mutationType !== MutationType.DELETE) {
        requestObject.data  = {
          ...data,
          mission_state: missionState
        }
      }
      return axiosClientService.axiosClient.request({...requestObject })
    },
    onSuccess: () => {
      queryClient.cancelQueries({ queryKey: missionControlKeys.all() });
      queryClient.invalidateQueries({ queryKey: missionControlKeys.all() });
    },
    onError: (error) => {
      console.error(error);
    }
  });

  const mutateMissionControl = ({ id, missionState = MissionState.PreFlight, mutationType, description, title }: MissionControlInput) => {
    const url = mutationType === MutationType.POST ? `flight_missions` : `flight_missions/${id}`;
    if (mutationType === MutationType.DELETE) {
      mutation.mutate({ mutationType, url });
    }  else {
      mutation.mutate({ missionState, mutationType, description, title, url });
    }

  };

  return {
    mutate: mutateMissionControl,
    isLoading: mutation.isLoading,
    isError: mutation.isError,
    error: mutation.error,
    isSuccess: mutation.isSuccess,
  };
}
