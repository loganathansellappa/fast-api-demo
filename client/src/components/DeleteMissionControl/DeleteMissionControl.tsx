import { MissionControlInput, useMissionControlUpdate } from '../../hooks/useMissionControlMutation.ts';
import { MutationType, showErrorToast, showSuccessToast } from '../../utils/HelperUtils.ts';
import { useCallback, useId } from 'react';

export type DeleteMissionControlProps = {
  onCancel: () => void;
  id: number;
}
export const DeleteMissionControl = ({ onCancel, id }: DeleteMissionControlProps ) => {
  const {  mutate, isError, isSuccess } = useMissionControlUpdate();
  const uId = useId();
  const onDeleteClick = useCallback(() => {
    const requestData: MissionControlInput = {
      mutationType: MutationType.DELETE,
      id,
    }
    mutate(requestData);
  }, [mutate, id]);


  if(isError) {
    showErrorToast('Mission Deletion Failed', `MissionDeletionFailed${uId}${id}`);
    onCancel();
  }

  if(isSuccess) {
    showSuccessToast('Mission Deletion Success', `MissionDeletionSuccess${uId}${id}`);
    onCancel();
  }

  return (
    <div className="flex flex-col">
      <div className="mb-12 pt-3">
        Are you sure? You can't undo this action afterwards.
      </div>
        <div className="flex justify-end">
          <button
            type="button"
            onClick={onCancel}
            className="cancel-delete mr-2 px-4 py-2 bg-gray-300 hover:bg-gray-400 text-gray-800 font-semibold rounded-md focus:outline-none focus:bg-gray-400"
          >
            Cancel
          </button>
          <button
            type="button"
            onClick={onDeleteClick}
            className="mission-delete px-4 py-2 bg-blue-800 hover:bg-blue-600 text-white font-semibold rounded-md focus:outline-none focus:bg-blue-600"
          >
            Delete
          </button>
        </div>
      </div>
  );
};
