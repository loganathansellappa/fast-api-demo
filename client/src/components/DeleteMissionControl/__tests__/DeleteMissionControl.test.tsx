import { render } from '@testing-library/react';
import * as Hooks from '../../../hooks/useMissionControlMutation.ts';
import { DeleteMissionControl, DeleteMissionControlProps } from '../DeleteMissionControl.tsx';

describe("DeleteMissionControl", () => {
  let element: { container: HTMLElement };
  const clickHandler = jest.fn();
  const submitHandler = jest.fn();

  beforeEach(() => {
    jest
      .spyOn(Hooks, "useMissionControlUpdate")
      .mockReturnValue(
        {
          mutate: submitHandler,
          isLoading: false,
          isError: false,
          error: false,
          isSuccess: false,
        }
      );
    const props: DeleteMissionControlProps = {
      onCancel: clickHandler,
      id: 1,
    };
    element = render(<DeleteMissionControl {...props} />);
  });

  test("renders DeleteMissionControl component", () => {
    expect(element.container.firstChild).toMatchSnapshot();
  });

  test("displays text", () => {
    expect(element.container.textContent).toContain("Are you sure? You can't undo this action afterwards.");
  });

  test("trigger click on cancel button click", () => {
    const cancelButton = element.container.getElementsByClassName("cancel-delete")[0] as HTMLElement;
    cancelButton.click();
    expect(clickHandler).toBeCalledTimes(1);
  });
});