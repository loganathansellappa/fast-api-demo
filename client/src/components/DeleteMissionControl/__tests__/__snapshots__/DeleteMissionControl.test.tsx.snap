// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeleteMissionControl renders DeleteMissionControl component 1`] = `
<div
  class="flex flex-col"
>
  <div
    class="mb-12 pt-3"
  >
    Are you sure? You can't undo this action afterwards.
  </div>
  <div
    class="flex justify-end"
  >
    <button
      class="cancel-delete mr-2 px-4 py-2 bg-gray-300 hover:bg-gray-400 text-gray-800 font-semibold rounded-md focus:outline-none focus:bg-gray-400"
      type="button"
    >
      Cancel
    </button>
    <button
      class="mission-delete px-4 py-2 bg-blue-800 hover:bg-blue-600 text-white font-semibold rounded-md focus:outline-none focus:bg-blue-600"
      type="button"
    >
      Delete
    </button>
  </div>
</div>
`;
